File: 02_selection_sort/python/01_selection_sort.py
# Finds the smallest value in an array
def findSmallest(arr):
  # Stores the smallest value
  # The function does the same thing as expression min(arr)
  smallest = arr[0]
  for i in arr[1:]:
    if smallest > i:
      smallest = i
  return smallest

# Sort array
def selectionSort(arr):
  newArr = []
  while arr:
      # Finds the smallest element in the array and adds it to the new array
      smallest = findSmallest(arr)
      newArr.append(smallest)
      arr.remove(smallest)
  return newArr

print(selectionSort([5, 3, 6, 2, 10]))

File: 02_selection_sort/javascript/01_selection_sort.js
'use strict';
/**
 * Finds the index of the element with the smallest value in the array
 * @param {Array} array Source array
 * @returns {number} Index of the element with the smallest value
 */
const findSmallest = (arr) => {
  let min = arr[0];
  let index = 0;
  for (let i = 1; i < arr.length; i++) {
    const el = arr[i];
    if (el >= min) continue;
    min = el;
    index = i;
  }
  return index;
};

/**
 * Sort array by increment
 * @param {Array} array Source array
 * @returns {Array} New sorted array
 */
const selectionSort = (arr) => {
  const size = arr.length;
  const result = new Array(size).fill(0);
  for (let i = 0; i < size; i++) {
    const smallestIndex = findSmallest(arr);
    const [curr] = arr.splice(smallestIndex, 1);
    result[i] = curr;
  }
  return result;
};

const sourceArray = [5, 3, 6, 2, 10];
const sourtedArray = selectionSort([5, 3, 6, 2, 10]);

console.log('Source array - ', sourceArray); // [5, 3, 6, 2, 10]
console.log('New sorted array - ', sourtedArray); // [2, 3, 5, 6, 10]
